[CI] AsyncSearchActionIT.testRestartAfterCompletion failing

`There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.`

**Build scan**:

https://gradle-enterprise.elastic.co/s/jfzgknxz6l7i6/tests/:x-pack:plugin:async-search:internalClusterTest/org.elasticsearch.xpack.search.AsyncSearchActionIT/testRestartAfterCompletion

**Repro line**:

```
./gradlew ':x-pack:plugin:async-search:internalClusterTest' --tests "org.elasticsearch.xpack.search.AsyncSearchActionIT.testRestartAfterCompletion" -Dtests.seed=D46DE9715D3991B7 -Dtests.locale=sr-CS -Dtests.timezone=America/Cambridge_Bay -Druntime.java=17
```

**Reproduces locally?**:

No

**Applicable branches**:

8.0

**Failure history**:

https://gradle-enterprise.elastic.co/scans/tests?search.relativeStartTime=P7D&search.timeZoneId=Europe/London&tests.container=org.elasticsearch.xpack.search.AsyncSearchActionIT&tests.sortField=FAILED&tests.test=testRestartAfterCompletion&tests.unstableOnly=true

**Failure excerpt**:

```
[2021-12-20T07:55:54,167][INFO ][o.e.x.s.AsyncSearchActionIT] [testRestartAfterCompletion] before test
[2021-12-20T07:55:54,319][INFO ][o.e.t.InternalTestCluster] [testRestartAfterCompletion] Restarting node [node_s4] 
[2021-12-20T07:55:54,320][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] stopping ...
[2021-12-20T07:55:54,327][INFO ][o.e.c.c.Coordinator      ] [node_s4] master node [{node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}] disconnected, restarting discovery
[2021-12-20T07:55:54,331][INFO ][o.e.t.ClusterConnectionManager] [node_s0] transport connection to [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}] closed by remote
[2021-12-20T07:55:54,331][INFO ][o.e.t.ClusterConnectionManager] [node_s1] transport connection to [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}] closed by remote
[2021-12-20T07:55:54,331][INFO ][o.e.t.ClusterConnectionManager] [node_s2] transport connection to [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}] closed by remote
[2021-12-20T07:55:54,332][INFO ][o.e.t.ClusterConnectionManager] [node_s5] transport connection to [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}] closed by remote
[2021-12-20T07:55:54,334][INFO ][o.e.t.ClusterConnectionManager] [node_s3] transport connection to [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}] closed by remote
[2021-12-20T07:55:54,340][INFO ][o.e.c.r.a.AllocationService] [node_s1] current.health="RED" message="Cluster health status changed from [GREEN] to [RED] (reason: [{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw} reason: disconnected])." previous.health="GREEN" reason="{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw} reason: disconnected"
[2021-12-20T07:55:54,341][WARN ][o.e.a.s.TransportClearScrollAction] [node_s4] Clear SC failed on node[{node_s3}{R5eoDjLrRsCq03g4uJMZPQ}{h6pCUtsAT0mNZfwwAYNdjQ}{127.0.0.1}{127.0.0.1:42310}{cdfhilrstw}{xpack.installed=true}]
org.elasticsearch.transport.NodeNotConnectedException: [node_s3][127.0.0.1:42310] Node not connected
	at org.elasticsearch.transport.ClusterConnectionManager.getConnection(ClusterConnectionManager.java:257) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.test.transport.StubbableConnectionManager.getConnection(StubbableConnectionManager.java:70) ~[framework-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService.getConnection(TransportService.java:761) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.SearchTransportService.getConnection(SearchTransportService.java:617) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.ClearScrollController.lambda$cleanScrollIds$4(ClearScrollController.java:113) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:136) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.SearchScrollAsyncAction.collectNodesAndRun(SearchScrollAsyncAction.java:116) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.ClearScrollController.cleanScrollIds(ClearScrollController.java:106) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.ClearScrollController.lambda$new$1(ClearScrollController.java:72) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.ClearScrollController.run(ClearScrollController.java:81) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.TransportClearScrollAction.doExecute(TransportClearScrollAction.java:39) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.TransportClearScrollAction.doExecute(TransportClearScrollAction.java:19) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:79) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:54) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:170) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:105) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:83) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:380) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.support.AbstractClient.clearScroll(AbstractClient.java:605) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.reindex.ClientScrollableHitSource.clearScroll(ClientScrollableHitSource.java:111) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.reindex.ScrollableHitSource.close(ScrollableHitSource.java:115) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.finishHim(AbstractAsyncBulkByScrollAction.java:559) [reindex-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.refreshAndFinish(AbstractAsyncBulkByScrollAction.java:521) [reindex-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.onBulkResponse(AbstractAsyncBulkByScrollAction.java:471) [reindex-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction$2.onResponse(AbstractAsyncBulkByScrollAction.java:420) [reindex-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction$2.onResponse(AbstractAsyncBulkByScrollAction.java:416) [reindex-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.Retry$RetryHandler.finishHim(Retry.java:168) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.Retry$RetryHandler.onResponse(Retry.java:112) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.Retry$RetryHandler.onResponse(Retry.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$0(NodeClient.java:107) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:176) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:170) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$RunBeforeActionListener.onResponse(ActionListener.java:389) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation$1.finishHim(TransportBulkAction.java:635) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation$1.onFailure(TransportBulkAction.java:630) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$1(NodeClient.java:114) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:185) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.finishAsFailed(TransportReplicationAction.java:1040) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$2.onClusterServiceClose(TransportReplicationAction.java:1025) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onClusterServiceClose(ClusterStateObserver.java:338) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onClose(ClusterStateObserver.java:245) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.addTimeoutListener(ClusterApplierService.java:254) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.ClusterStateObserver.waitForNextChange(ClusterStateObserver.java:177) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.ClusterStateObserver.waitForNextChange(ClusterStateObserver.java:118) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.ClusterStateObserver.waitForNextChange(ClusterStateObserver.java:110) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.retry(TransportReplicationAction.java:1017) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleException(TransportReplicationAction.java:996) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService$4.handleException(TransportService.java:719) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1343) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService$5.doRun(TransportService.java:882) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:220) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:856) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:740) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:663) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.performAction(TransportReplicationAction.java:968) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.performLocalAction(TransportReplicationAction.java:909) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.doRun(TransportReplicationAction.java:890) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction.runReroutePhase(TransportReplicationAction.java:226) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction.doExecute(TransportReplicationAction.java:221) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction.doExecute(TransportReplicationAction.java:84) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:79) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:54) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:170) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:105) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation.doRun(TransportBulkAction.java:605) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction.executeBulk(TransportBulkAction.java:757) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction.doInternalExecute(TransportBulkAction.java:280) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$1.doRun(TransportBulkAction.java:200) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:775) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
[2021-12-20T07:55:54,348][INFO ][o.e.c.s.MasterService    ] [node_s1] node-left[{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw} reason: disconnected], term: 1, version: 26, delta: removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}
[2021-12-20T07:55:54,385][INFO ][o.e.c.s.ClusterApplierService] [node_s0] removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}, term: 1, version: 26, reason: ApplyCommitRequest{term=1, version=26, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:54,395][INFO ][o.e.c.s.ClusterApplierService] [node_s2] removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}, term: 1, version: 26, reason: ApplyCommitRequest{term=1, version=26, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:54,452][INFO ][o.e.c.s.ClusterApplierService] [node_s5] removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}, term: 1, version: 26, reason: ApplyCommitRequest{term=1, version=26, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:54,460][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][17] start check index
[2021-12-20T07:55:54,462][INFO ][o.e.c.s.ClusterApplierService] [node_s3] removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}, term: 1, version: 26, reason: ApplyCommitRequest{term=1, version=26, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:54,485][INFO ][o.e.c.s.ClusterApplierService] [node_s1] removed {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{GiLwkfzJR9-ljuFX8nH8ng}{127.0.0.1}{127.0.0.1:38956}{cdfhilrstw}}, term: 1, version: 26, reason: Publication{term=1, version=26}
[2021-12-20T07:55:54,523][INFO ][o.e.c.r.DelayedAllocationService] [node_s1] scheduling reroute for delayed shards in [59.8s] (1 delayed shards)
[2021-12-20T07:55:54,525][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][17] end check index
[2021-12-20T07:55:54,537][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][8] start check index
[2021-12-20T07:55:54,536][INFO ][o.e.t.s.M.Listener       ] [node_s4] [.async-search][0] start check index
[2021-12-20T07:55:54,618][INFO ][o.e.i.s.IndexShard       ] [node_s3] [.async-search][0] primary-replica resync completed with 0 operations
[2021-12-20T07:55:54,651][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][8] end check index
[2021-12-20T07:55:54,695][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][11] start check index
[2021-12-20T07:55:54,841][INFO ][o.e.t.s.M.Listener       ] [node_s4] [.async-search][0] end check index
[2021-12-20T07:55:54,893][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][11] end check index
[2021-12-20T07:55:54,932][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][2] start check index
[2021-12-20T07:55:54,967][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][2] end check index
[2021-12-20T07:55:54,983][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][14] start check index
[2021-12-20T07:55:55,057][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][14] end check index
[2021-12-20T07:55:55,069][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][5] start check index
[2021-12-20T07:55:55,132][INFO ][o.e.t.s.M.Listener       ] [node_s4] [test-async][5] end check index
[2021-12-20T07:55:55,140][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] stopped
[2021-12-20T07:55:55,140][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] closing ...
[2021-12-20T07:55:55,166][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] closed
[2021-12-20T07:55:55,206][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] version[8.0.0-SNAPSHOT], pid[277163], build[unknown/unknown/f83e7d3b605891c2571c1718c12b85006c053d57/2021-12-20T14:31:19.347918747Z], OS[Linux/3.10.0-1160.49.1.el7.x86_64/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/17.0.1/17.0.1+12-39]
[2021-12-20T07:55:55,206][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] JVM home [/var/lib/jenkins/.java/openjdk-17.0.1+12-linux]
[2021-12-20T07:55:55,206][WARN ][o.e.d.n.Node             ] [testRestartAfterCompletion] data_stream.dataset="deprecation.elasticsearch" data_stream.namespace="default" data_stream.type="logs" elasticsearch.event.category="other" event.code="no-jdk" message="no-jdk distributions that do not bundle a JDK are deprecated and will be removed in a future release"
[2021-12-20T07:55:55,207][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] JVM arguments [-Dfile.encoding=UTF8, -Dsun.jnu.encoding=UTF8, -Des.scripting.update.ctx_in_params=false, -Des.search.rewrite_sort=true, -Des.transport.cname_in_publish_address=true, -Dgradle.dist.lib=/var/lib/jenkins/.gradle/wrapper/dists/gradle-7.3.2-all/4h7zmfyirw3l9vsxqbzv7bhr8/gradle-7.3.2/lib, -Dgradle.user.home=/var/lib/jenkins/.gradle, -Dgradle.worker.jar=/var/lib/jenkins/.gradle/caches/7.3.2/workerMain/gradle-worker.jar, -Dio.netty.noKeySetOptimization=true, -Dio.netty.noUnsafe=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Djava.awt.headless=true, -Djava.locale.providers=SPI,COMPAT, -Djna.nosys=true, -Dorg.gradle.internal.worker.tmpdir=/dev/shm/elastic+elasticsearch+8.0+periodic+java-matrix/x-pack/plugin/async-search/build/tmp/internalClusterTest/work, -Dorg.gradle.native=false, -Dtests.artifact=async-search, -Dtests.gradle=true, -Dtests.logger.level=WARN, -Dtests.security.manager=true, -Dtests.seed=D46DE9715D3991B7, -Dtests.task=:x-pack:plugin:async-search:internalClusterTest, --illegal-access=deny, --add-opens=java.base/java.security.cert=ALL-UNNAMED, --add-opens=java.base/java.nio.channels=ALL-UNNAMED, --add-opens=java.base/java.net=ALL-UNNAMED, --add-opens=java.base/javax.net.ssl=ALL-UNNAMED, --add-opens=java.base/java.nio.file=ALL-UNNAMED, --add-opens=java.base/java.time=ALL-UNNAMED, --add-opens=java.management/java.lang.management=ALL-UNNAMED, -XX:+HeapDumpOnOutOfMemoryError, -esa, -XX:+UseG1GC, -XX:HeapDumpPath=/dev/shm/elastic+elasticsearch+8.0+periodic+java-matrix/x-pack/plugin/async-search/build/heapdump, --add-opens=java.base/java.util=ALL-UNNAMED, --add-opens=java.base/java.lang=ALL-UNNAMED, -Xms512m, -Xmx512m, -Dfile.encoding=UTF-8, -Djava.io.tmpdir=/dev/shm/elastic+elasticsearch+8.0+periodic+java-matrix/x-pack/plugin/async-search/build/testrun/internalClusterTest/temp, -Duser.country=US, -Duser.language=en, -Duser.variant, -ea]
[2021-12-20T07:55:55,207][WARN ][o.e.n.Node               ] [testRestartAfterCompletion] version [8.0.0-SNAPSHOT] is a pre-release version of Elasticsearch and is not suitable for production
[2021-12-20T07:55:55,208][WARN ][o.e.d.n.Node             ] [testRestartAfterCompletion] data_stream.dataset="deprecation.elasticsearch" data_stream.namespace="default" data_stream.type="logs" elasticsearch.event.category="settings" event.code="shared-data-path" message="setting [path.shared_data] is deprecated and will be removed in a future release"
[2021-12-20T07:55:55,208][WARN ][o.e.d.n.Node             ] [testRestartAfterCompletion] data_stream.dataset="deprecation.elasticsearch" data_stream.namespace="default" data_stream.type="logs" elasticsearch.event.category="settings" event.code="multiple-data-paths" message="Configuring multiple [path.data] paths is deprecated. Use RAID or other system level features for utilizing multiple disks. This feature will be removed in a future release."
[2021-12-20T07:55:55,220][WARN ][o.e.d.n.Node             ] [testRestartAfterCompletion] data_stream.dataset="deprecation.elasticsearch" data_stream.namespace="default" data_stream.type="logs" elasticsearch.event.category="settings" event.code="multiple-data-paths-list" message="Configuring [path.data] with a list is deprecated. Instead specify as a string value."
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] no modules loaded
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.node.NodeMocksPlugin]
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.reindex.ReindexPlugin]
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.script.MockScriptService$TestPlugin]
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.search.MockSearchService$TestPlugin]
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.ESIntegTestCase$AssertActionNamePlugin]
[2021-12-20T07:55:55,221][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.ESIntegTestCase$TestSeedPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.MockHttpTransport$TestPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.TestGeoShapeFieldMapperPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.store.MockFSIndexStore$TestPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.test.transport.MockTransportService$TestPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.transport.nio.MockNioTransportPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.xpack.async.AsyncResultsIndexPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.xpack.core.LocalStateCompositeXPackPlugin]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.xpack.search.AsyncSearch]
[2021-12-20T07:55:55,222][INFO ][o.e.p.PluginsService     ] [testRestartAfterCompletion] loaded plugin [org.elasticsearch.xpack.search.AsyncSearchIntegTestCase$SearchTestPlugin]
[2021-12-20T07:55:55,272][INFO ][o.e.e.NodeEnvironment    ] [testRestartAfterCompletion] using [3] data paths, mounts [[/dev/shm (tmpfs)]], net usable_space [25.7gb], net total_space [47.1gb], types [tmpfs]
[2021-12-20T07:55:55,273][INFO ][o.e.e.NodeEnvironment    ] [testRestartAfterCompletion] heap size [512mb], compressed ordinary object pointers [true]
[2021-12-20T07:55:55,308][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] node name [node_s4], node ID [vFaZsxN8Qk6MR2ZSLrMgtg], cluster name [SUITE-TEST_WORKER_VM=[675]-CLUSTER_SEED=[167947222111106922]-HASH=[16775D259F5]-cluster], roles [data_cold, transform, ml, data_content, data_hot, remote_cluster_client, ingest, data, data_frozen, data_warm]
[2021-12-20T07:55:55,383][INFO ][o.e.d.DiscoveryModule    ] [testRestartAfterCompletion] using discovery type [zen] and seed hosts providers [settings, file]
[2021-12-20T07:55:55,415][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] initialized
[2021-12-20T07:55:55,416][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] starting ...
[2021-12-20T07:55:55,433][INFO ][o.e.t.TransportService   ] [testRestartAfterCompletion] publish_address {127.0.0.1:33279}, bound_addresses {[::1]:40136}, {127.0.0.1:33279}
[2021-12-20T07:55:55,750][INFO ][o.e.c.c.Coordinator      ] [testRestartAfterCompletion] cluster UUID [APvrsWTCSVqVmK1m_Rl9Qg]
[2021-12-20T07:55:55,873][INFO ][o.e.c.s.MasterService    ] [node_s1] node-join[{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw} join existing leader], term: 1, version: 28, delta: added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}
[2021-12-20T07:55:55,892][INFO ][o.e.c.s.ClusterApplierService] [node_s2] added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}, term: 1, version: 28, reason: ApplyCommitRequest{term=1, version=28, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:55,896][INFO ][o.e.c.s.ClusterApplierService] [node_s3] added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}, term: 1, version: 28, reason: ApplyCommitRequest{term=1, version=28, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:55,907][INFO ][o.e.c.s.ClusterApplierService] [node_s5] added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}, term: 1, version: 28, reason: ApplyCommitRequest{term=1, version=28, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:55,908][INFO ][o.e.c.s.ClusterApplierService] [node_s0] added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}, term: 1, version: 28, reason: ApplyCommitRequest{term=1, version=28, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:55,909][INFO ][o.e.c.s.ClusterApplierService] [node_s4] master node changed {previous [], current [{node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}]}, added {{node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}, {node_s5}{pcWxIKDTSlqiwlIIivlX_g}{EVqgNLMARSKVVcKpiou1wg}{127.0.0.1}{127.0.0.1:35013}{cdfhilrstw}, {node_s0}{xCCCuUQaRYGhEx2pq-xzsw}{7whpwHoOStGiI4U8zrY6NQ}{127.0.0.1}{127.0.0.1:37047}{ilmrt}, {node_s2}{5v-EZfODTdqF45P4P8Be9A}{OT3tO7rzT4mLg-xTzkmBtw}{127.0.0.1}{127.0.0.1:44752}{ilmrt}, {node_s3}{R5eoDjLrRsCq03g4uJMZPQ}{h6pCUtsAT0mNZfwwAYNdjQ}{127.0.0.1}{127.0.0.1:42310}{cdfhilrstw}}, term: 1, version: 28, reason: ApplyCommitRequest{term=1, version=28, sourceNode={node_s1}{ZQfeFCugR0yej8EUTMuLgQ}{6ds6lNXOR6S_i40gHq1cCw}{127.0.0.1}{127.0.0.1:35769}{ilmrt}{xpack.installed=true}}
[2021-12-20T07:55:55,998][INFO ][o.e.l.LicenseService     ] [node_s4] license [d21ee172-405f-4892-a058-65b3c3e13f4f] mode [basic] - valid
[2021-12-20T07:55:56,000][INFO ][o.e.n.Node               ] [testRestartAfterCompletion] started
[2021-12-20T07:55:56,007][INFO ][o.e.c.s.ClusterApplierService] [node_s1] added {{node_s4}{vFaZsxN8Qk6MR2ZSLrMgtg}{mE_EiU_9T1C5ZbZzLU8UuQ}{127.0.0.1}{127.0.0.1:33279}{cdfhilrstw}}, term: 1, version: 28, reason: Publication{term=1, version=28}
[2021-12-20T07:55:57,270][INFO ][o.e.c.r.a.AllocationService] [node_s1] current.health="YELLOW" message="Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[test-async][2]]])." previous.health="RED" reason="shards started [[test-async][2]]"
[2021-12-20T07:55:57,385][WARN ][o.e.c.r.a.AllocationService] [node_s1] [.async-search][0] marking unavailable shards as stale: [pbYmp3lxR4iJtJzu-7CW6A]
[2021-12-20T07:55:57,551][INFO ][o.e.c.r.a.AllocationService] [node_s1] current.health="GREEN" message="Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[.async-search][0]]])." previous.health="YELLOW" reason="shards started [[.async-search][0]]"
```

Here's my understanding of what happened: in `testRestartAfterCompletion`, one of the nodes is restarted. By bad luck, the background async search clean-up task happened to run around the same time as the restart. This task uses delete-by-query under the hood, which in turn uses scrolls, and its scroll context failed to be cleaned up:
```
[2021-12-20T07:55:54,341][WARN ][o.e.a.s.TransportClearScrollAction] [node_s4] Clear SC failed on node[{node_s3}{R5eoDjLrRsCq03g4uJMZPQ}{h6pCUtsAT0mNZfwwAYNdjQ}{127.0.0.1}{127.0.0.1:42310}{cdfhilrstw}{xpack.installed=true}]
org.elasticsearch.transport.NodeNotConnectedException: [node_s3][127.0.0.1:42310] Node not connected
  at org.elasticsearch.transport.ClusterConnectionManager.getConnection(ClusterConnectionManager.java:257) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
```

This makes all subsequent tests fail because there are unexpected open contexts:
```
    java.lang.AssertionError: There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.
        at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:56)
```

Usually the async search clean-up tasks runs very infrequently, but in this suite we've set it to run every 100ms to test the clean-up functionality. Perhaps we could split this suite in two: one where the clean-up task never runs (so there's no risk of it running during a restart), and one that explicitly tests the clean-up functionality where it runs every 100ms?

--- a/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
+++ b/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
@@ -27,7 +27,6 @@ import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.plugins.SearchPlugin;
 import org.elasticsearch.reindex.ReindexPlugin;
 import org.elasticsearch.rest.RestStatus;
-import org.elasticsearch.search.MockSearchService;
 import org.elasticsearch.search.aggregations.bucket.filter.InternalFilter;
 import org.elasticsearch.search.builder.PointInTimeBuilder;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
@@ -153,7 +152,7 @@ public abstract class AsyncSearchIntegTestCase extends ESIntegTestCase {
 
         // Temporarily stop garbage collection, making sure to wait for any in-flight tasks to complete
         stopMaintenanceService();
-        assertBusy(MockSearchService::assertNoInFlightContext);
+        ensureAllSearchContextsReleased();
 
         internalCluster().restartNode(node.getName(), new InternalTestCluster.RestartCallback() {
         });

--- a/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
+++ b/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
@@ -27,6 +27,7 @@ import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.plugins.SearchPlugin;
 import org.elasticsearch.reindex.ReindexPlugin;
 import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.MockSearchService;
 import org.elasticsearch.search.aggregations.bucket.filter.InternalFilter;
 import org.elasticsearch.search.builder.PointInTimeBuilder;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
@@ -149,7 +150,11 @@ public abstract class AsyncSearchIntegTestCase extends ESIntegTestCase {
         AsyncExecutionId searchId = AsyncExecutionId.decode(id);
         final ClusterStateResponse clusterState = client().admin().cluster().prepareState().clear().setNodes(true).get();
         DiscoveryNode node = clusterState.getState().nodes().get(searchId.getTaskId().getNodeId());
+
+        // Temporarily stop garbage collection, making sure to wait for any in-flight tasks to complete
         stopMaintenanceService();
+        assertBusy(MockSearchService::assertNoInFlightContext);
+
         internalCluster().restartNode(node.getName(), new InternalTestCluster.RestartCallback() {
         });
         startMaintenanceService();

